design-doc.md

```diff
@@ -0,0 +1,128 @@
+
+The quantum storage unit of the stream is an **entry**.
+
+Entries can be appended to streams using **publishers** and scanned with **listeners**.
+
+
+## Entries
+
+Streams are strongly typed, and within a stream, each entry has the type specified at stream declaration time. All entries within one stream have the same type. The type can, and often is, polymorphic.
+
+Implementation-wise, entries of polymorphic type are `std::unique_ptr<>`-s of `SomeBaseType`: the user they call their virtual functions or have individual instances of `[const] SomeBaseType&` dispatched via other RTTI mechanisms.
+
+The type of entries in the stream is defined at stream declaration time. When serializing/deserealizing streams to/from disk, a basic check of the signature of storage schema is performed to verify that no data corruption may occur.
+
+Polymorphic types can be extended with more derived types, but it is the user's responsibilty that no binary attempts to deserialize entries of the type not supported by this particular version. (`TODO(dkorolev): See how strict we can be with this in V1.`)
+
+In order to qualify for a stream type, it should be guaranteed that an entry of this type can be serialized and de-serialized, and that the very same entry will be constructed as a result of the serialization/de-serialization cycle. In other words, stream entries should be agnostic with respect to serialization. This is a requirement for dropping strict guarantees on the number of serializations that take place internally.
+
+Implemenation-wise, Sherlock uses Cereal, and we use JSON and binary formats.
+
+## Consistency
+
+Each entry of a stream is assigned an **index**: an `uint64_t` value equal to the number of entries preceeding the one being considered, effectively, a 0-based index.
+
+Additionally, a stream defines the **order key type**. One can think of the order key as the timestamp.
+
+Order keys often are indeed timestamps, represented in certain way, for example, Unix epoch time in milliseconds.
+
+It is the user's responsibility to ensure that order keys of consecutive entries follow a strictly increasing order. (`TODO(dkorolev): I am still debating whether "non-decreasing order" is worth it here, leaning towards "no".`).
+
+This guarantee enables `TailProduce`: A paradigm for real-time data crunching algorithms based on `producers` accepting data joined at runtime from multiple streams on their respectiveve order keys.
+
+Implementation-wise, the mechanism to extract order keys from entries is a member function that can be templated or overloaded. Thus, there are no constraints on how many various values can be used as order keys per entry type. Keep in mind though, that if the entry type is polymorphic, in order for certain order key type to be supported it is required for a getter of the order key of this type to be a (virtual) function of the top-level base class of this polymorphic type.
```

Comment:
From this section it's not obvious to me, how the concepts **order key type**, **order key** and **various values can be used as order keys** are related.